src/Matrix.c

```diff
@@ -0,0 +1,237 @@
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+
+
+#include "ErrorCode.h"
+#include "Matrix.h"
+
+typedef struct Matrix{
+	size_t height;
+	size_t width;
+	double* ptr;
+} Matrix;
+
+
+// check if the allocate failed
+#define CHECK_MATRIX_ALLOC(matrix)                           \
+    if ((matrix) == NULL){                                   \
+        return ERROR_MALLOC_FAILED;                          \
+	}
+
+// check if the given matrix pointer and data pointer are valid
+#define CHECK_MATRIX_MEMORY(matrix)                          \
+    if ((matrix) == NULL || (matrix)->ptr == NULL) {         \
+        return ERROR_MATRIX_IS_NULL;                         \
+	}
+
+// check if a matrix function succeseed and stop the program if its not
+#define CHECK_MATRIX_CALL(call)                              \
```

Comment:
This macro is quite dangerous as it might redeclare variables, read about ## and `__LINE__`